### PR TITLE
prov/verbs: Reformat GID type output

### PR DESCRIPTION
**Pull request by lradeke**
The Verbs provider defines a fabric name based on the GID subnet prefix returned from ibv_query_gid(), which is currently hardcoded to 0 and therefore only returns the subnet prefix for GID[0] . This results in fi_info only returning only "IB-0xfe80000000000000" as the fabric attribute output, regardless of which device it queries. This implies that all devices queried reside on the same fabric, and may result in applications using the wrong endpoints during communication.

With multiple interfaces going to multiple fabrics, the subnet prefixes should be set differently than default. Modifications to the verbs_info.c file replace the hardcoded 0 with vrb_gl_data.gid_idx and introduce a switch case to add a RoCEv1 or v2 prefix where appropriate. The fi_info fabric attribute output for RoCE devices is now more easily differentiated:

    fabric: RoCEv1-fe80000000000000 (RoCEv1 w/IPv6)
    fabric: RoCEv2-fe80000000000000 (RoCEv2 w/IPv6)
    fabric: RoCEv1-0 (RoCEv1 w/IPv4)
    fabric: RoCEv2-0 (RoCEv2 w/IPv4)

Signed-off-by: Leena Radeke <leena.radeke@hpe.com>

### CODE REVIEW

File: prov/verbs/src/verbs_info.c [hunk 1]
```diff
@@ -1811,5 +1844,6 @@ int vrb_getinfo(uint32_t version, const char *node, const char *service,
 	if (!ret || ret == -FI_ENOMEM || ret == -FI_ENODEV)
 		return ret;
 	else
-		return -FI_ENODATA;
+                return -FI_ENODATA;
```
Comment (shefty) [inline prov/verbs/src/verbs_info.c:1847]: unrelated change is converting tabs into spaces for some reason.

File: prov/verbs/src/verbs_info.c [hunk 2]
```diff
@@ -791,21 +794,51 @@ static int vrb_alloc_info(struct ibv_context *ctx, struct fi_info **info,
 
 	switch (ctx->device->transport_type) {
 	case IBV_TRANSPORT_IB:
-		if (ibv_query_gid(ctx, 1, 0, &gid)) {
+		if (ibv_query_gid(ctx, 1, vrb_gl_data.gid_idx, &gid)) {
```
Comment (shefty) [inline prov/verbs/src/verbs_info.c:797]: Support for gid_idx should be a separate patch

File: prov/verbs/src/fi_verbs.h [hunk 3]
```diff
@@ -484,6 +484,17 @@ int vrb_srq_context(struct fid_domain *domain, struct fi_rx_attr *attr,
 int vrb_domain_xrc_init(struct vrb_domain *domain);
 int vrb_domain_xrc_cleanup(struct vrb_domain *domain);
 
+enum ibv_gid_type {
+	IBV_GID_TYPE_IB_ROCE_V1,
+	IBV_GID_TYPE_ROCE_V2,
+};
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.h:490]: This enum should be within an #if check, and not duplicated if already defined by verbs.h.  Also, IBV_GID_TYPE_IB should be the first value, so that we don't define an enum value with the same name, but a different value from verbs.h.

File: prov/verbs/src/fi_verbs.h [hunk 4]
```diff
@@ -484,6 +484,17 @@ int vrb_srq_context(struct fid_domain *domain, struct fi_rx_attr *attr,
...
+
+#if !HAVE_DECL_IBV_QUERY_GID_TYPE && HAVE_IBV_QUERY_GID_TYPE
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.h:492]: A single define here should be sufficient.  E.g. #if VERBS_HAVE_GID_TYPE

File: prov/verbs/configure.m4 [hunk 5]
```diff
@@ -57,6 +57,20 @@ AC_DEFUN([FI_VERBS_CONFIGURE],[
...
+		AC_CHECK_LIB([ibverbs], [ibv_query_gid_type],[],[],[])
+		AC_CHECK_FUNCS([ibv_query_gid_type],[],[],
+			[[#include <infiniband/drivers.h>]])
+		AC_CHECK_DECL([ibv_query_gid_type],
+			[HAVE_DECL_IBV_QUERY_GID_TYPE=1],[],
+			[[#include <infiniband/drivers.h>]])
+		])
+	AC_DEFINE_UNQUOTED([HAVE_DECL_IBV_QUERY_GID_TYPE],
+		[$HAVE_DECL_IBV_QUERY_GID_TYPE],
+		[Whether verbs have ibv_query_gid_type() support or not])
```
Comment (shefty) [inline prov/verbs/configure.m4:72]: Please match other tests done for verbs, including matching the naming convention (e.g. VERBS_HAVE_GID_TYPE)

File: prov/verbs/src/fi_verbs.h [hunk 6]
```diff
@@ -484,6 +484,17 @@ int vrb_srq_context(struct fid_domain *domain, struct fi_rx_attr *attr,
...
+	extern int *ibv_query_gid_type(struct ibv_context *context,
+					uint8_t port_num, unsigned int index,
+					enum ibv_gid_type *gid_type);
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.h:495]: We need some sort of implementation to avoid an unresolved symbol error.  Please test the build with an older version of verbs where this function does not exist.

File: prov/verbs/src/verbs_info.c [hunk 7]
```diff
@@ -791,21 +794,51 @@ static int vrb_alloc_info(struct ibv_context *ctx, struct fi_info **info,
...
 			goto err;
 		}
 
-		name_len = strlen(VERBS_IB_PREFIX) + INET6_ADDRSTRLEN;
-		if (!(fi->fabric_attr->name = calloc(1, name_len + 1))) {
-			ret = -FI_ENOMEM;
+		if (ibv_query_gid_type(ctx, 1, vrb_gl_data.gid_idx, &gid_type)) {
+			VERBS_INFO_ERRNO(FI_LOG_FABRIC,
+					 "ibv_query_gid_type", errno);
+			ret = -errno;
```
Comment (shefty) [inline prov/verbs/src/verbs_info.c:807]: Rather than failing, we can check the link type and assume either IB or RoCE.  Worst case, we shouldn't fail, but keep the existing behavior and just report as IB.

File: prov/verbs/src/verbs_info.c [hunk 8]
```diff
@@ -791,21 +794,51 @@ static int vrb_alloc_info(struct ibv_context *ctx, struct fi_info **info,
...
+				 VERBS_ROCEV2_PREFIX "%" PRIx64,
+				 be64toh(gid.global.subnet_prefix));
+			break;
+		default:
+			name_len = strlen(VERBS_IB_PREFIX) + INET6_ADDRSTRLEN;
+			if (!(fi->fabric_attr->name = calloc(1, name_len + 1))) {
+				ret = -FI_ENOMEM;
+				goto err;
+			}
+
```
Comment (shefty) [inline prov/verbs/src/verbs_info.c:839]: This allocates space for the name, then does nothing?

File: prov/verbs/src/verbs_info.c [hunk 9]
```diff
@@ -791,21 +794,51 @@ static int vrb_alloc_info(struct ibv_context *ctx, struct fi_info **info,
...
+		case IBV_GID_TYPE_ROCE_V2:
+			name_len = strlen(VERBS_ROCEV2_PREFIX) + INET6_ADDRSTRLEN;
+			if (!(fi->fabric_attr->name = calloc(1, name_len + 1))) {
+				ret = -FI_ENOMEM;
+				goto err;
+			}
+
+			snprintf(fi->fabric_attr->name, name_len,
+				 VERBS_ROCEV2_PREFIX "%" PRIx64,
+				 be64toh(gid.global.subnet_prefix));
```
Comment (shefty) [inline prov/verbs/src/verbs_info.c:831]: Is there no meaningful value for RoCE GIDs that we can record?  The description shows that this just returns 0.

**Review by shefty — CHANGES_REQUESTED** [review]

File: prov/verbs/src/verbs_info.c (hunk 9 above)
Comment (lradeke) [reply prov/verbs/src/verbs_info.c:831]: This was actually the reason I posted the linked issue #6365. The idea was to have the IPv4 RoCE devices to list their respective IPv4 addresses after the prefixes. Converting the IPv6-formatted GID to IPv4 and having that print isn't a good choice, as more than 1 device can share a single GID index and the GIDs are currently being queried via GID index. The idea of reassigning the verbs_fabric_attr value to something specific to the fabric rather than to the provider has been floated -- instead of passing back the global info for the Verbs provider to return the specific GID, tailor the fabric_attr field to contain something fabric-specific and populate the information into that per-fabric data structure. Thoughts towards how to implement this? Or different ideas entirely?

Comment (shefty) [reply prov/verbs/src/verbs_info.c:831]: If we can use the IP address, I would be fine with that.  The tcp/udp providers do this, so there's code available to discover IP addresses and convert them.  I'm okay even removing the 'RoCEv2' portion of the string in that case, since ideally the protocol conveys that same information.

Do we know how the RoCEv2 GID value is formed?  Is the IP address buried in that, or does that contain some MAC address?

Comment (jswaro) [reply prov/verbs/src/verbs_info.c:831]: IIRC, the GID value is formed from the IPv4 or IPv6 address associated with the gid. It's the reason why GID 0 is often 0xfe800000 (IPv6 autoconf address). For IPv6, the assumption to date has been they always have a 64 bit prefix. For IPv4, we didn't really encounter the gid parsing because of the default assumption of GID 0.

Comment (shefty) [reply prov/verbs/src/verbs_info.c:831]: If that's the case, then trying to use the IP address should basically result in the same value, right?  I.e. we would still get the IPv6 autoconf address.

File: prov/verbs/src/fi_verbs.h (hunk 6 above)
Comment (shefty) [reply prov/verbs/src/fi_verbs.h:496]: The function is being declared (extern shouldn't be needed), but it's never defined.

File: prov/verbs/src/verbs_info.c [hunk 10]
```diff
@@ -798,14 +801,46 @@ static int vrb_alloc_info(struct ibv_context *ctx, struct fi_info **info,
...
+		}
```
Comment (shefty) [inline prov/verbs/src/verbs_info.c:842]: The above code can be simplified.  It's nearly the same in each case.  E.g.
```
char *prefix;

if (ibv_query_gid_type(...)) {
    gid_type = IB; /* assume IB as default for backward compatibility */
}

switch (gid_type) {
case ROCE_V1:
    prefix = ROCEV1_PREFIX;
    break;
case ROCE_V2:
    prefix = ROCEV2_PREFIX;
    break;
default:
    prefix = IB_PREFIX;
    break;
}

name_len = strlen(prefix) + INET6;
name = calloc(1, name_len + 1);
snprintf(name, name_len, "%s%" PRIx64, prefix, subnet_prefix);
```

File: prov/verbs/src/verbs_info.c (hunk 9 above)
Comment (jswaro) [reply prov/verbs/src/verbs_info.c:830]: Yes, I think it should work.

**Comment by shefty** [comment]
No activity for months.  If still needed, please update with changes and (re) open a PR.  CI issues would need to be addressed.
